lib: Supporting building for `wasm32-wasip1`

Earthmark opened the pull request:
It's really strange not putting a description of changes in this message.

This is my first set of JJ managed changes, I hope this works.

Earthmark commented on lib/src/local_working_copy.rs:
```diff
@@ -106,7 +106,7 @@ use crate::working_copy::WorkingCopyStateError;
 #[cfg(unix)]
 type FileExecutableFlag = bool;
-#[cfg(windows)]
+#[cfg(any(windows, target_os = "wasi"))]
```
I can comment this in the repo if wanted, but wasi doesn't have a concept of file permission flags right now https://github.com/WebAssembly/wasi-filesystem/blob/main/wit/types.wit

thoughtpolice replied:
Yes, please do leave a comment. Is there a ticket or some kind of ideas for this? It would be enough to just leave a comment to a relevant ticket i.e. `// wasi symlink support: https://github.com/webassembly/wasi-filesystem/issue/...`

Earthmark replied:
I haven't found any workstreams trying to add executable flag support for wasi.

I would assume executable files would be added in a separate interface. I'm pretty sure wasi doesn't have the ability to execute another process right now in the current spec, marking a file as executable wouldn't be meaningful.

I'll add a comment as well for this line, if I do run into an issue link I'll append that.

martinvonz reviewed (approved):
Thanks. I suppose it would make sense to add this build to CI. Do you want to follow up with that?

martinvonz reviewed (commented):
If anyone here knows more about WebAssembly (more than ~nothing, that is), please review this. Otherwise I think we can merge this in a day or so.

PhilipMetzger commented:
Can you please adjust the commit message to `lib: Supporting building for wasm32-wasip1`, see https://github.com/martinvonz/jj/blob/main/docs/contributing.md#code-reviews. And I'd like to know the motivation behind supporting it, if there are any.  And maybe it'd be nice to coordinate the changes with #4478, since that PR is a major rework for `lib/src/local_working_copy.rs`?

Earthmark commented:
Oh something went wrong, I'll see what happened locally and get that resolved.

thoughtpolice reviewed (commented):
Seems fine, but let's add a build matrix entry too. I suspect we'll probably need to remove git support for that to work? I would be surprised if `libgit2` and the needed deps all build cleanly.

If we could actually build the CLI without git support, we would in theory throw away most of the external deps, and be able to write a WASI backend that used the browser filesystem APIs. Then we could host a version of JJ in the browser as an interactive tutorial. :)

Earthmark commented:
> Thanks. I suppose it would make sense to add this build to CI. Do you want to follow up with that?

Step added, it should verify `lib` builds for wasi.

Earthmark commented:
This should probably wait until https://github.com/martinvonz/jj/pull/4478 goes through, I feel that change is likely a higher precedence and it'd be easy to rebuild these changes.

Earthmark commented:
> Can you please adjust the commit message to `lib: Supporting building for wasm32-wasip1`, see https://github.com/martinvonz/jj/blob/main/docs/contributing.md#code-reviews. And I'd like to know the motivation behind supporting it, if there are any. And maybe it'd be nice to coordinate the changes with #4478, since that PR is a major rework for `lib/src/local_working_copy.rs`?

Commit message adjusted.

The use case for this is building out very trivial vfs-wasi support to https://github.com/RobertBaruch/dergwasm to represent entities in the game world as entries in a filesystem, and running jj as a wasm executable to version objects inside the VR environment, from inside the environment itself.

@thoughtpolice 's use case of a web tool on an example filesystem is a much more realizeable use case. 

I said this in another comment, but I feel #4478 should probably go through first, as it's a much more complex change.

Earthmark commented:
Turns out the executable bit of every file got set on accident.

My main machine is windows, but I worked on this repo in a linux dev container with the folder mounted into the container. I'm pretty sure it exposes mounted windows files as 755, which got committed to the repo by the dev container. In my initial change I ran jj from the windows env, for the follow up I ran it in the container itself, which is why the issue only came up in the follow up.

Earthmark committed:
lib: Supporting building for wasm32-wasip1

This is adapting the FS apis in `lib` for wasi (excluding git support).

The compilation command for building this is:
`cargo +nightly build --no-default-features --target wasm32-wasip1`

* Nightly is needed for `std::os::wasi::fs::symlink_path`, if symlink support isn't needed then this can be non-nightly.
* No default features is needed to not compile gitoxide, as that doesn't yet support wasi.

This does not adapt the CLI, just lib.

martinvonz commented:
Oops :) Sounds like #4478 will be useful for you too then.

PhilipMetzger commented:
> > Can you please adjust the commit message to `lib: Supporting building for wasm32-wasip1`, see https://github.com/martinvonz/jj/blob/main/docs/contributing.md#code-reviews. And I'd like to know the motivation behind supporting it, if there are any. And maybe it'd be nice to coordinate the changes with #4478, since that PR is a major rework for `lib/src/local_working_copy.rs`?
> 
> Commit message adjusted.

Thanks. 

> The use case for this is building out very trivial vfs-wasi support to https://github.com/RobertBaruch/dergwasm to represent entities in the game world as entries in a filesystem, and running jj as a wasm executable to version objects inside the VR environment, from inside the environment itself.

That sounds like a really interesting idea, although I think that just having the auto-snapshotting mechanism could be enough for you.

Thank you for enabling the support for it.

Earthmark commented:
The ability to restructure changes is very useful as well. I'm happy to talk about it over discord if you're curious!

pksunkara commented:
Superseded by #9849
